pkg/trustedresources/verify.go

```diff
@@ -100,18 +89,99 @@ func VerifyPipeline(ctx context.Context, pipelineObj v1beta1.PipelineObject, k8s
+			if err := verifyInterface(resource, verifier, signature); err == nil {
+				return nil
+			}
+		}
+		return fmt.Errorf("resource %s in namespace %s fails verification", resource.GetName(), resource.GetNamespace())
+	}
+
+	if len(policies) == 0 {
+		return errorEmptyVerificationConfig
+	}
+
+	matchedPolicies := matchedPolicies(policies, source)
+	if len(matchedPolicies) == 0 {
+		return fmt.Errorf("no matching policies are found for resource: %s against source: %s. %w", resource.GetName(), source, errorNoMatchedPolicies)
+	}
+
+	verifiedPolicies, err := verifiedPolicies(ctx, resource, k8s, signature, matchedPolicies)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to verify the resource with matched policies: %w", err)
+	}
+
+	// make sure all matched policies pass verification
+	if verifiedPolicies == len(matchedPolicies) {
+		return nil
 	}
+	return fmt.Errorf("resource %s in namespace %s fails verification %w", resource.GetName(), resource.GetNamespace(), ErrorResourceVerificationFailed)
+
+}
 
-	for _, verifier := range verifiers {
-		if err := VerifyInterface(pipeline, verifier, signature); err == nil {
-			return nil
+// matchedPolicies return all matched VerificationPolicy by checking if the Pattern is matching source.
+func matchedPolicies(policies []*v1alpha1.VerificationPolicy, source string) []*v1alpha1.VerificationPolicy {
+	matchedPolicies := []*v1alpha1.VerificationPolicy{}
+	for _, vp := range policies {
+		for _, r := range vp.Spec.Resources {
+			matching, _ := regexp.MatchString(r.Pattern, source)
+			if matching {
+				matchedPolicies = append(matchedPolicies, vp)
+				break
+			}
+		}
+	}
+	return matchedPolicies
+}
+
+// verifiedPolicies loop all policies to get the count of verified policies. For each policy
+// the resource passes if there is one public key can verify the signature.
+func verifiedPolicies(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, policies []*v1alpha1.VerificationPolicy) (int, error) {
+	verifiedPolicies := 0
+	for _, p := range policies {
+		verifiers, err := verifiersFromPolicy(ctx, k8s, p)
+		if err != nil {
+			return verifiedPolicies, fmt.Errorf("failed to get verifiers from policy: %w", err)
 		}
+		for _, verifier := range verifiers {
+			// if one of the verifier passes verification, then this policy passes verification
+			if err := verifyInterface(resource, verifier, signature); err == nil {
+				verifiedPolicies++
+				break
+			}
+		}
+	}
+	return verifiedPolicies, nil
+}
+
+// verifyInterface get the checksum of json marshalled object and verify it.
+func verifyInterface(obj interface{}, verifier signature.Verifier, signature []byte) error {
+	ts, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+
+	h := sha256.New()
+	h.Write(ts)
+
+	if err := verifier.VerifySignature(bytes.NewReader(signature), bytes.NewReader(h.Sum(nil))); err != nil {
+		return errors.Wrap(ErrorResourceVerificationFailed, err.Error())
```

Comment:
why do some functions return wrapped errors with `errors.Wrap` and some don't? im thinking we'd want to be consistent about which ones do to make sure anyone using exported functions could depend on expecting to get a wrapped error or not

Reply:
fixed